src/DiagonalHessianApproximation.jl

```diff
@@ -0,0 +1,145 @@
+using LinearAlgebra
+using LinearOperators
+
+function mulSquareOpDiagonal!(res, d, v, α, β::T) where {T <: Real}
+  if β == zero(T)
+    res .= α .* d .* v
+  else
+    res .= α .* d .* v .+ β .* res
+  end
+end
+
+abstract type AbstractDiagonalQuasiNewtonOperator{T} <: AbstractLinearOperator{T} end
+
+"""
+Implementation of the diagonal quasi-Newton approximation described in
+
+Andrei, N. 
+A diagonal quasi-Newton updating method for unconstrained optimization. 
+https://doi.org/10.1007/s11075-018-0562-7
+"""
+
+mutable struct DiagonalQN{T <: Real, I <: Integer, V <: AbstractVector{T}} <: AbstractDiagonalQuasiNewtonOperator{T} 
+  d::V # Diagonal of the operator
+  Bs::V # Preallocate a vector used in push! function
+  nrow::I
+  ncol::I
+  symmetric::Bool
+  hermitian::Bool
+  prod!
+  tprod!
+  ctprod!
+  nprod::I
+  ntprod::I
+  nctprod::I
+  args5::Bool
+  use_prod5!::Bool # true for 5-args mul! and for composite operators created with operators that use the 3-args mul!
+  Mv5::V
+  Mtu5::V
+  allocated5::Bool # true for 5-args mul!, false for 3-args mul! until the vectors are allocated
+end
+
+DiagonalQN(d::AbstractVector{T}) where {T <: Real} = 
+  DiagonalQN(
+    d,
+    d,
+    length(d),
+    length(d),
+    true, 
+    true,  
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β), 
+    0,
+    0,
+    0,
+    true,
+    true,
+    typeof(d)(undef,0),
+    typeof(d)(undef,0),
+    true)
+
+# update function
+# s = x_{k+1} - x_k
+# y = ∇f(x_{k+1}) - ∇f(x_k)
+function push!(
+  B::DiagonalQN{T,I,V},
+  s::V,
+  y::V
+  ) where {T <: Real, I <: Integer, V <: AbstractVector{T}}
+  trA2 = zero(T)
+  for i in eachindex(s)
+    trA2 += s[i]^4
+  end
+  sT_s = dot(s,s)
+  sT_y = dot(s,y)
+  mul!(B.Bs,B,s)
+  sT_B_s = dot(s,B.Bs)
```

Comment:
In fact, do we really need the temporary arrays `Bs`, `Mv5` and `Mtu5` for diagonal operators? You could write
```julia
mul!(B.Bs, B, s)
s^T_B_s = dot(s, B.Bs)
```
as
```julia
s^T_B_s = sum(s[i]^2 * B.d[i] for i ∈ eachindex(s))
```
Similarly, when a user calls `mul5!(res, B, v, α, β)`, your implementation doesn't require temporary storage. I think that's why you defined `Mv5` and `Mtu5` as `typeof(d)(0)` in the constructor (which should be `similar(d, 0)`.

So actually, I think you could remove `Bs`, `Mv5` and `Mtu5` from the structs. They are not required to be there.